jsonrpcbase/main.py

```diff
@@ -165,122 +150,92 @@ def call_py(self, jsondata):
                         responds.append(respond)
-
                 if responds:
                     return responds
-
                 # Nothing to respond.
                 return None
             else:
                 # empty dict, list or wrong type
                 raise InvalidRequestError
-
         except InvalidRequestError as e:
             return self._get_err(e, request['id'])
+        except InvalidIDType as err:
+            return self._invalid_id_type_response(err, request)
         except JSONRPCError as e:
             return self._get_err(e, request['id'], request['jsonrpc'])
+        except jsonschema.exceptions.ValidationError as e:
+            return self._get_jsonschema_err(e, request['id'], request['jsonrpc'])
+
+    def _invalid_id_type_response(self, err, request):
+        """
+        Response for an invalid `id` field type
+        """
+        err = {
+            'message': err.message,
+            'code': err.code,
+        }
+        resp = {'id': None, 'error': err}
+        return resp
+
+    def _get_jsonschema_err(self, err, id=None, jsonrpc=DEFAULT_JSONRPC):
+        """
+        Returns an error message for a jsonschema validation error.
+        """
+        resp = {'id': id}
+        self._fill_ver(jsonrpc, resp)
+        resp['error'] = {
+            'message': err.message,
+            'code': -32602,
+        }
+        return resp
 
     def _get_err(self, e, id=None, jsonrpc=DEFAULT_JSONRPC):
         """
         Returns jsonrpc error message.
         """
         # Do not respond to notifications when the request is valid.
-        if not id and not isinstance(e, ParseError) and not isinstance(e, InvalidRequestError):
+        if id is None and not isinstance(e, ParseError) and not isinstance(e, InvalidRequestError):
             return None
-
         respond = {'id': id}
-
-        if isinstance(jsonrpc, int):
-            # v1.0 requires result to exist always.
-            # No error codes are defined in v1.0 so only use the message.
-            if jsonrpc == 10:
-                respond['result'] = None
-                respond['error'] = e.dumps()['message']
-            else:
-                self._fill_ver(jsonrpc, respond)
-                respond['error'] = e.dumps()
-        else:
-            respond['jsonrpc'] = jsonrpc
-            respond['error'] = e.dumps()
-
+        self._fill_ver(jsonrpc, respond)
+        respond['error'] = e.dumps()
         return respond
 
-    def _fill_ver(self, iver, respond):
+    def _fill_ver(self, ver, respond):
         """
-        Fills version information to the respond from the internal integer version.
+        Fills version information to the respond from the internal tuple version.
         """
-        if iver == 20:
+        if ver == (2, 0):
             respond['jsonrpc'] = '2.0'
-        if iver == 11:
+        elif ver == (1, 1):
             respond['version'] = '1.1'
 
-    def _vargs(self, f):
```

Comment:
This validation stuff is removed and replaced with the jsonschema

Reply:
Nice to get rid of the homegrown stuff and replace with a nicer way!